sw/device/lib/otbn.c

```diff
@@ -0,0 +1,162 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "otbn.h"
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_otbn.h"
+#include "sw/device/lib/runtime/hart.h"
+
+static bool is_func_ptr(const otbn_t *ctx, const otbn_ptr_t *ptr) {
+  return ptr->ptr_in_cpu_addr_space >= ctx->app->imem_start &&
+         ptr->ptr_in_cpu_addr_space <= ctx->app->imem_end;
+}
+
+static uint32_t func_ptr_in_otbn_addr_space(const otbn_t *ctx,
+                                            const otbn_ptr_t *ptr) {
+  return ptr->ptr_in_cpu_addr_space - ctx->app->imem_start;
+}
+
+static bool is_data_ptr(const otbn_t *ctx, const otbn_ptr_t *ptr) {
+  return ptr->ptr_in_cpu_addr_space >= ctx->app->dmem_start &&
+         ptr->ptr_in_cpu_addr_space <= ctx->app->dmem_end;
+}
+
+static uint32_t data_ptr_in_otbn_addr_space(const otbn_t *ctx,
+                                            const otbn_ptr_t *ptr) {
+  return ptr->ptr_in_cpu_addr_space - ctx->app->dmem_start;
+}
```

Comment:
Could you give these functions comments explaining what they do?

Reply:
I folded the is_data_ptr() and data_ptr_in_otbn_addr_space() into a single function to have the pre-condition check together with the operation and added a doxygen comment. I'm not convinced it's significantly better, as we now have a out param to deal with. Win some, loose some.